File: src/riskmeasures/simulation.py
import numpy as np
from riskmeasures.utils import validate_int, validate_positive

def simulate_data(N : int,
                  beta_0 : float = 1,
                  beta_1 : float = 2,
                  x_sigma : float = 1,
                  epsilon_sigma : float = 1,
                  seed : int | None = None
                  ) -> np.ndarray:
    """
    Simulate data according to the linear regression model:
        y_i = beta_0 + beta_1 * x_i + epsilon_i,  i = 1, ..., N

    where:
        x_i ~ N(0, x_sigma^2)
        epsilon_i ~ N(0, epsilon_sigma^2)
    
    Returns a 2D array where each row is an (x, y) pair.

    Parameters
    ----------
    N : int
        Number of observations to generate.
    beta_0 : float
        Intercept of the regression model.
    beta_1 : float
        Slope of the regression model.
    x_sigma : float
        Standard deviation of the x values.
    epsilon_sigma : float
        Standard deviation of the error terms.
    seed : int, optional
        Random seed for reproducibility.

    Returns
    -------
    data : np.ndarray
        A 2D array where each row is an (x, y) pair.
    """
    validate_int("N", N, min_value=2)
    validate_positive("x_sigma", x_sigma)
    validate_positive("epsilon_sigma", epsilon_sigma)

    rng = np.random.default_rng(seed)
    xs = rng.normal(0, x_sigma, N)
    epsilons = rng.normal(0, epsilon_sigma, N)
    ys = beta_0 + beta_1 * xs + epsilons
    data = np.column_stack((xs, ys))
    return data

File: src/riskmeasures-R/simulation.R
#Simulate data according to the linear regression model:
# y_i = beta_0 + beta_1 * x_i + epsilon_i
#where x_i ~ N(0, x_sigma^2) and epsilon_i ~ N(0, epsilon_sigma^2)

#source(file.path(dirname(sys.frame(1)$ofile), "utils.R"))

simulate_data <- function(N,
                          beta_0 = 1,
                          beta_1 = 2,
                          x_sigma = 1,
                          epsilon_sigma = 1,
                          seed = NULL) {
#' @param N Number of observations
#' @param beta_0 Intercept
#' @param beta_1 Slope
#' @param x_sigma Standard deviation of x values
#' @param epsilon_sigma Standard deviation of error terms
#' @param seed Optional random seed

  if (!is.numeric(N) || length(N) != 1 || is.na(N) || N %% 1 != 0) {
    stop("N must be an integer")
  }
  if (N < 2) {
    stop(sprintf("N must be >= 2, got %d", N))
  }
  if (!is.numeric(x_sigma) || length(x_sigma) != 1 || !is.finite(x_sigma) || x_sigma <= 0) {
    stop("x_sigma must be a positive finite number")
  }
  if (!is.numeric(epsilon_sigma) || length(epsilon_sigma) != 1 || !is.finite(epsilon_sigma) || epsilon_sigma <= 0) {
    stop("epsilon_sigma must be a positive finite number")
  }
  if (!is.null(seed)) {
    validate_and_set_seed(seed)
  }

  xs <- rnorm(N, mean = 0, sd = x_sigma)
  epsilons <- rnorm(N, mean = 0, sd = epsilon_sigma)
  ys <- beta_0 + beta_1 * xs + epsilons
  
  cbind(x = xs, y = ys)
}
